### DISCUSSION TITLE
Eclipse CDT 10.0 and Eclipse CDT 11.2 compatibility

### DISCUSSION THREAD

**Discussion by mmx85** [eclipse-cdt/cdt#550] [Q&A] [answered]
Hello.
I have 3 Eclipse CDT compatibility questions:
1. Are projects created for Eclipse CDT 10.0 and Eclipse CDT 11.2 compatible? I mean can I create a project using Eclipse CDT 10.0 and then continue working with it using Eclipse CDT 11.2 and vice-versa?
2. Are .cproject files of CDT 10.0 and Eclipse CDT 11.2 compatible? They have some differences like 
a) "keepEnvironmentInBuildfile" argument added for "builder" xml element in Eclipse CDT 11.2 
b) "storageModule" xml element added in Eclipse CDT 11.2 
3. For options like ilg.gnuarmeclipse.managedbuild.cross..... new argument "name" appeared in .cproject file after update of GNU MCU Eclipse 4.7.2 plug-in to Eclipse CDT plugin 6.3.2. Does this affects compatibility between GNU MCU Eclipse 4.7.2 plug-in and Eclipse CDT plugin 6.3.2

For now I noticed one incompatibility - when project is created using Eclipse CDT 10.0 and built using external builder then optimal number of build jobs can be determined (user can choose this option) and added automatically as an argument, say "-j8", but for Eclipse CDT 11.2 you need to set number of build jobs manually only. Automated determination works only for internal builder. This is a little bit inconvenient.

**Comment by jonahgraham** [comment, marked as answer]
Short answer is we provide backwards compatibility guarantee only, no forward guarantee. i.e. it is a (serious!) bug if you open a project created in an old version of CDT in the latest release and it doesn't work. But there is no guarantee the other way around. That said, in practice it will probably work fine as the CDT format has been stable since well before CDT 10.0

So specifically:

> Are projects created for Eclipse CDT 10.0 and Eclipse CDT 11.2 compatible? I mean can I create a project using Eclipse CDT 10.0 and then continue working with it using Eclipse CDT 11.2 and vice-versa?

Yes and no - CDT 10.0 projects can be opened in 11.2, but no guarantee on vice-versa.

> Are .cproject files of CDT 10.0 and Eclipse CDT 11.2 compatible? They have some differences like
> a) "keepEnvironmentInBuildfile" argument added for "builder" xml element in Eclipse CDT 11.2

That setting has existed since before CDT 10.0, so not sure why you are seeing a difference.

> b) "storageModule" xml element added in Eclipse CDT 11.2

That setting has existed since before CDT 10.0, so not sure why you are seeing a difference.

> For options like ilg.gnuarmeclipse.managedbuild.cross..... new argument "name" appeared in .cproject file after update of GNU MCU Eclipse 4.7.2 plug-in to Eclipse CDT plugin 6.3.2. Does this affects compatibility between GNU MCU Eclipse 4.7.2 plug-in and Eclipse CDT plugin 6.3.2

Whether GNU MCU projects are forwards compatible with Eclipse Embed CDT projects is a question for https://github.com/eclipse-embed-cdt/eclipse-plugins/discussions - but I suspect the same answer applies.

>  say "-j8", but for Eclipse CDT 11.2 you need to set number of build jobs manually only.

That sounds like a bug related to fixing how [make arguments](https://github.com/eclipse-cdt/cdt/blob/main/NewAndNoteworthy/CDT-10.3.md#build-command-arguments-moved-to-behavior-tab) are handled some time ago (but since CDT 10.0). The .cproject file format wasn't supposed to have changed, just the UI so it should work. FWIW A quick check with CDT 11.3 external builder shows that `parallelBuildOn="true" parallelizationNumber="optimal"` works and that *should* work with CDT 10.0 too. If not your experience, please raise a bug.

**Reply by mmx85** [reply]
Thank you for reply.

**Reply by jonahgraham** [reply]
> Whether GNU MCU projects are forwards compatible with Eclipse Embed CDT projects is a question for https://github.com/eclipse-embed-cdt/eclipse-plugins/discussions - but I suspect the same answer applies.

Cross reference to https://github.com/eclipse-embed-cdt/eclipse-plugins/discussions/589 which is @mmx85's question on Eclipse Embed CDT
